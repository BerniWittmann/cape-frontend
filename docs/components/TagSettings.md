# TagSettings

A Settings Tab to edit the tags

## Methods

<!-- @vuese:TagSettings:methods:start -->
|Method|Description|Parameters|
|---|---|---|
|updateLayoutTable|Update the Table Layout|-|

<!-- @vuese:TagSettings:methods:end -->


