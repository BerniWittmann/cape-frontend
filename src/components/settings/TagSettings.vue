<template>
  <el-row type="flex" justify="center">
    <el-col :span="8">
      <el-row>
        <el-col>
          <tag class="tag-row__tag" :tag="tag" v-for="tag in tags" :key="tag.id" closable @close="deleteTag(tag)"></tag>
        </el-col>
      </el-row>
      <br>

      <el-row>
        <el-col>
          <tag :tag="tag" v-if="tag.name && tag.name.length > 0"></tag>
        </el-col>
      </el-row>

      <el-form ref="tagForm" :model="tag" label-position="top" :rules="rules">
        <el-form-item :label="$t('settings.tag.name')" prop="name">
          <el-input v-model="tag.name"></el-input>
        </el-form-item>
        <el-form-item :label="$t('settings.tag.color')" prop="color">
          <el-color-picker v-model="tag.color"></el-color-picker>
        </el-form-item>
        <el-form-item>
          <el-button type="success" @click.native="submit">{{ $t('settings.tag.submit') }}</el-button>
          <el-button type="danger" plain @click.native="reset">{{ $t('settings.tag.reset') }}</el-button>
        </el-form-item>
      </el-form>
    </el-col>
  </el-row>
</template>

<script>
/* ============
 * Tag Settings Component
 * ============
 *
 * A component to edit the Tags
 */

import { DEFAULT_TAG_COLOR } from '@/utils/constants'
import Tag from '@/models/tag'
import tagService from '@/services/tag'

import TagComponent from '@/components/Tag.vue'

export default {
  components: {
    Tag: TagComponent
  },

  data() {
    return {
      tag: {
        name: undefined,
        color: DEFAULT_TAG_COLOR
      },
      rules: {
        name: [
          { required: true, message: this.$t('settings.tag.validation.name.required'), trigger: 'blur' }
        ],
        color: [
          { required: true, message: this.$t('settings.tag.validation.color.required'), trigger: 'blur' }
        ]
      }
    }
  },

  methods: {
    submit() {
      this.$refs.tagForm.validate((valid) => {
        if (valid) {
          const tag = new Tag(this.tag)
          tagService.create(tag).then(() => {
            this.reset()
          })
        }
      })
    },

    reset() {
      this.tag = {
        name: undefined,
        color: DEFAULT_TAG_COLOR
      }
      this.$refs.tagForm.resetFields()
    },

    deleteTag(tag) {
      tagService.remove(tag)
    }
  },

  computed: {
    tags() {
      return this.$store.state.tag.tags
    }
  }
}

</script>

<style scoped lang="scss">
.tag-row__tag {
  margin-bottom: 5px;
}
</style>
