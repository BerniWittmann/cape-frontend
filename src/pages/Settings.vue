<template>
  <v-layout>
    <div class="page">
      <h3>{{ $t('settings.title') }}</h3>
      <h4>{{ $t('settings.tags') }}</h4>
      <tag-settings></tag-settings>
    </div>
  </v-layout>
</template>

<script>
/* ============
 * Settings Page
 * ============
 *
 * The Settings Page
 */
import DefaultLayout from '@/layouts/Default.vue'
import TagSettings from '@/components/settings/TagSettings.vue'

export default {
  components: {
    VLayout: DefaultLayout,
    TagSettings
  }
}
</script>

<style scoped lang="scss">
.page {
  text-align: center;
}
</style>
