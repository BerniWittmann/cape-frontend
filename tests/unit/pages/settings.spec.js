import { shallowMount, mount } from '@vue/test-utils'
import { i18n } from '../setupPlugins'

import Settings from '@/pages/Settings.vue'
import TagSettings from '@/components/settings/TagSettings.vue'
import EmptySlotComponent from '../EmptySlotComponent.vue'

console.error = jest.fn()

// mock the random function to get always the same id for working snapshots
const mockMath = Object.create(global.Math)
mockMath.random = () => 0.5
global.Math = mockMath

describe('Pages', () => {
  describe('Settings', () => {
    it('renders', () => {
      const wrapper = shallowMount(Settings, {
        i18n,
        stubs: {
          'v-layout': EmptySlotComponent
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('updates the context type table layout on tab click', () => {
      Settings.updateContextType = jest.fn()
      let store = {
        state: {
          contextType: {
            contextTypes: [{
              id: '142',
              name: 'My first ContextType',
              icon: 'fa-heart'
            }]
          },
          tag: {
            tags: [{
              id: '142',
              name: 'My Tag',
              color: '#FF0000'
            }]
          }
        }
      }

      let cmp = mount(Settings, {
        i18n,
        mocks: {
          $store: store
          // sync: false
        }
      })

      expect(cmp.html()).toMatchSnapshot()
      const tabs = cmp.findAll('.el-tab-pane')
      expect(tabs.length).toBe(3)

      let tab = {
        label: 'settings.tags'
      }
      let m = jest.fn()

      cmp.vm.$refs.tagSettings.updateLayoutTable = m
      cmp.find('.el-tabs').vm.$emit('tab-click', tab)
      expect(m).toHaveBeenCalled()

      tab.label = 'settings.context_types'
      cmp.vm.$refs.contextTypeSettings.updateLayoutTable = m
      cmp.find('.el-tabs').vm.$emit('tab-click', tab)
      expect(m).toHaveBeenCalled()
    })
  })
})
