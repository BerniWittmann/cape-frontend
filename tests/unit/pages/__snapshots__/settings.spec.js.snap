// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages Settings renders 1`] = `
<div>
  <div class="page">
    <h3>settings.title</h3>
    <eltabs-stub type="border-card" tabposition="top">
      <eltabpane-stub label="settings.tags">
        <tagsettings-stub></tagsettings-stub>
      </eltabpane-stub>
      <eltabpane-stub label="settings.context_types">
        <contexttypesettings-stub></contexttypesettings-stub>
      </eltabpane-stub>
    </eltabs-stub>
  </div>
</div>
`;
